### PR TITLE
Cross compile to Scala 2.12 and 2.13

### PR DESCRIPTION
**Pull request by renatocaval**
References #243

### CODE REVIEW

File: akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala [hunk 1]
```diff
@@ -82,10 +82,7 @@ object ProjectionId {
    * @return an [[java.util.Set]] of [[ProjectionId]]s
    */
   def of(name: String, keys: java.util.Set[String]): java.util.Set[ProjectionId] =
-    keys
-      .stream()
-      .map(key => new ProjectionId(name, key))
-      .collect(Collectors.toSet())
+    keys.asScala.map { key: String => new ProjectionId(name, key) }.asJava
```
Comment (renatocaval) [inline akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala:85]: I'm wondering is we need this method. We can leave it for now and reassess when we go back to the ShardedDaemonProcess integration.

File: akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala [hunk 2]
```diff
@@ -298,7 +300,11 @@ private[akka] abstract class InternalProjectionState[Offset, Envelope](
         .futureSource(
           handlerStrategy.lifecycle
             .tryStart()
-            .flatMap(_ => sourceProvider.source(() => readOffsets())))
+            .flatMap { _ =>
+              sourceProvider
+                .source(() => readOffsets())
+                .map(_.mapMaterializedValue(_ => NotUsed))
```
Comment (renatocaval) [inline akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala:306]: We need to map it to `NotUsed` over and over again. Wondering if we should instead change the signature to `Source[Envolope, NotUsed]`, it's now `Source[Envolope, _]`

File: akka-projection-core/src/main/scala/akka/projection/internal/OffsetSerialization.scala [hunk 3]
```diff
@@ -75,11 +75,14 @@ import akka.projection.StringKey
...
       case mrg: MergeableOffset[_, _] =>
-        MultipleOffsets(mrg.entries.map {
+        val list = mrg.entries.map {
           case (key: MergeableKey, innerOffset) =>
             toStorageRepresentation(ProjectionId(id.name, key.surrogateKey), innerOffset, mergeable = true)
               .asInstanceOf[SingleOffset]
-        }.toSeq)
+        }.toList
+
+        MultipleOffsets(list)
```
Comment (renatocaval) [inline akka-projection-core/src/main/scala/akka/projection/internal/OffsetSerialization.scala:84]: map.toSeq is mutable in 2.12. 

I pull it out from the constructor to reduce some letter-soup. Hard to parse for the brackets when all together.

File: akka-projection-core/src/main/scala/akka/projection/internal/SourceProviderAdapter.scala [hunk 4]
```diff
@@ -30,9 +30,9 @@ import akka.stream.scaladsl.Source
     // it _should_ not be used for the blocking operation of getting offsets themselves
     val ec = akka.dispatch.ExecutionContexts.parasitic
     val offsetAdapter = new Supplier[CompletionStage[Optional[Offset]]] {
-      override def get(): CompletionStage[Optional[Offset]] = offset().map(_.toJava)(ec).asJava
+      override def get(): CompletionStage[Optional[Offset]] = offset().map(_.asJava)(ec).toJava
```
Comment (renatocaval) [inline akka-projection-core/src/main/scala/akka/projection/internal/SourceProviderAdapter.scala:33]: This is astonishing! 

I never realised that Option moved from `toJava` to `asJava` while Future from `asJava` to `toJava`. Let's wait for 2.14, we may need to revert it again.

File: akka-projection-kafka/src/main/scala/akka/projection/kafka/internal/KafkaSourceProviderImpl.scala [hunk 5]
```diff
@@ -91,12 +91,11 @@ import org.apache.kafka.common.TopicPartition
...
-              case (topicPartitionKey, offset) =>
-                val tp = topicPartitionKey.tp
-                if (assignedTps.contains(tp)) Map(tp -> offset)
-                else Map.empty
-            })
+            val filteredMap = groupOffsets.entries.collect {
+              case (topicPartitionKey, offset) if assignedTps.contains(topicPartitionKey.tp) =>
+                (topicPartitionKey.tp -> offset)
+            }
+            Future.successful(filteredMap)
```
Comment (renatocaval) [inline akka-projection-kafka/src/main/scala/akka/projection/kafka/internal/KafkaSourceProviderImpl.scala:98]: Here the inference got broken, something with CanBuild stuff. It turns out that we don't need to `flatMap` and return `Map.empty` for the cases we want to discard. `Map.collect` is all what we need.

File: project/Common.scala [hunk 6]
```diff
@@ -36,7 +36,7 @@ object Common extends AutoPlugin {
     projectInfoVersion := (if (isSnapshot.value) "snapshot" else version.value),
     crossVersion := CrossVersion.binary,
     crossScalaVersions := Dependencies.ScalaVersions,
-    scalaVersion := Dependencies.Scala213,
+    scalaVersion := Dependencies.Scala212,
```
Comment (renatocaval) [inline project/Common.scala:39]: default to 2.12 so we can spot issue faster

File: examples/src/test/scala/docs/cassandra/CassandraProjectionDocExample.scala [hunk 7]
```diff
@@ -193,7 +194,7 @@ object CassandraProjectionDocExample {
     ShardedDaemonProcess(system).init[ProjectionBehavior.Command](
       name = "shopping-carts",
       numberOfInstances = ShoppingCart.tags.size,
-      behaviorFactory = n => ProjectionBehavior(projection(ShoppingCart.tags(n))),
+      behaviorFactory = (n: Int) => ProjectionBehavior(projection(ShoppingCart.tags(n))),
```
Comment (renatocaval) [inline examples/src/test/scala/docs/cassandra/CassandraProjectionDocExample.scala:197]: no idea why this is need, but compiler said so

File: akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickOffsetStore.scala [hunk 8]
```diff
@@ -61,8 +62,10 @@ import slick.jdbc.JdbcProfile
   def saveOffset[Offset](projectionId: ProjectionId, offset: Offset): slick.dbio.DBIO[_] = {
     val now: Instant = Instant.now(clock)
     toStorageRepresentation(projectionId, offset) match {
-      case offset: SingleOffset  => newRow(offset, now)
-      case MultipleOffsets(reps) => DBIO.sequence(reps.map(rep => newRow(rep, now)))
+      case offset: SingleOffset => newRow(offset, now)
+      case MultipleOffsets(reps) =>
+        val actions = reps.map(rep => newRow(rep, now))
+        DBIO.sequence(actions)
```
Comment (renatocaval) [inline akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickOffsetStore.scala:68]: Amazing, but just assigning it to a val was enough to fix the inference! 🤷‍♂️

File: akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala [hunk 9]
```diff
@@ -4,10 +4,10 @@
 
 package akka.projection
 
-import java.util.stream.Collectors
-
 import scala.collection.immutable
 
+import akka.util.ccompat.JavaConverters._
```
Comment (patriknw) [inline akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala:9]: it's somewhat unfortunate to depend on Akka ineternals, but ok.
That was what I meant with that it will be difficult to avoid warnings for both 2.12 and 2.13, but ok let's use the akka ccompat

File: akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala (hunk 1 above)
Comment (patriknw) [reply akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala:85]: why change it to Scala?

File: akka-projection-core/src/main/scala/akka/projection/internal/HandlerRecoveryImpl.scala [hunk 10]
```diff
@@ -171,7 +171,7 @@ import akka.projection.StatusObserver
                 retries + 1,
                 exception)
             }
-            retried.recoverWith(_ => futDone)
+            retried.recoverWith { case _ => futDone }
```
Comment (patriknw) [inline akka-projection-core/src/main/scala/akka/projection/internal/HandlerRecoveryImpl.scala:174]: is that not allowed in 2.13?

File: akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala (hunk 2 above)
Comment (patriknw) [reply akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala:306]: sounds good to change, since we don't use it

File: examples/src/test/scala/docs/cassandra/CassandraProjectionDocExample.scala [hunk 11]
```diff
@@ -4,6 +4,7 @@
 
 package docs.cassandra
 
+import akka.actor.typed.scaladsl._
```
Comment (patriknw) [inline examples/src/test/scala/docs/cassandra/CassandraProjectionDocExample.scala:7]: be more specific, LoggerOps

**Comment by renatocaval** [comment]
Rebased after merging #249

File: akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala (hunk 1 above)
Comment (renatocaval) [reply akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala:85]: I will give it another try. The call `map(key => new ProjectionId(name, key))` wasn't compiling in Scala 2.12. I don't remember what was the reason.

File: akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala (hunk 2 above)
Comment (renatocaval) [reply akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala:314]: Ok, I will add to this PR. It will impact all the current PRs, but better to have all there at once.

File: akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala (hunk 9 above)
Comment (renatocaval) [reply akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala:9]: I thought you suggested that at some point.

File: akka-projection-core/src/main/scala/akka/projection/internal/HandlerRecoveryImpl.scala (hunk 10 above)
Comment (renatocaval) [reply akka-projection-core/src/main/scala/akka/projection/internal/HandlerRecoveryImpl.scala:174]: You man 2.12. 
Yes, this is not allowed.

File: akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala (hunk 1 above)
Comment (renatocaval) [reply akka-projection-core/src/main/scala/akka/projection/ProjectionId.scala:85]: this is the error we get in 2.12

```scala
[error]  found   : java.util.function.Function[String,akka.projection.ProjectionId]
[error]  required: java.util.function.Function[_ >: String, _ <: R]
[error]       .map(key => new ProjectionId(name, key))

```

I tried to solve this by using 
`import scala.compat.java8.FunctionConverters._` and different encoding alternatives to see if the compiler would understand what I wanted, but nothing worked out. So I kind of give up and decided to use scala collections instead

File: akka-projection-cassandra/src/main/scala/akka/projection/cassandra/internal/CassandraProjectionImpl.scala [hunk 12]
```diff
@@ -159,7 +159,7 @@ import akka.stream.scaladsl.Source
    * This is mainly intended to be used by the TestKit allowing it to attach a TestSink to it.
    */
   @InternalApi
-  override private[projection] def mappedSource()(implicit system: ActorSystem[_]): Source[Done, _] = {
+  override private[projection] def mappedSource()(implicit system: ActorSystem[_]): Source[Done, Future[Done]] = {
```
Comment (renatocaval) [inline akka-projection-cassandra/src/main/scala/akka/projection/cassandra/internal/CassandraProjectionImpl.scala:162]: As discussed [here](https://github.com/akka/akka-projection/pull/250#discussion_r442285497), the Mat type needs to fixed. At the end, it's not `NotUsed` but `Future[Done]`. That's because of the last transformation we do in `InternalProjectionState.mappedSource`.

File: akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala [hunk 13]
```diff
@@ -350,19 +356,20 @@ private[akka] abstract class InternalProjectionState[Offset, Envelope](
...
+          .andThen {
+            case Success(_) =>
+              statusObserver.stopped(projectionId)
+            case Failure(AbortProjectionException) =>
+              statusObserver.stopped(projectionId) // no restart
+            case Failure(exc) =>
+              Try(statusObserver.stopped(projectionId))
+              statusObserver.failed(projectionId, exc)
+          }
+      }
```
Comment (renatocaval) [inline akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala:373]: It's here, here we make the final `Source` to mat to `Future[Done]`. This wasn't visible before because the signature was `Source[Done, _]`

File: akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala (hunk 2 above)
Comment (renatocaval) [reply akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala:314]: So, there is two things here: 

1. the type returned by the SourceProvider which is still `Source[Done, _]` and we need to map it to something because of Scala 2.12. 
1. the type returned by `mappedSource` is now `Future[Done]`. This one we want it stable because is internal stuff and it will alway be `Future[Done]`.

I will do another iteration for change the type of the SourceProvider to `Source[Done, NotUsed]`,
